clang/test/SemaOpenCLCXX/address-space-lambda.cl

```diff
@@ -31,8 +31,8 @@ __kernel void test_qual() {
 //CHECK: |-CXXMethodDecl {{.*}} constexpr operator() 'void () const __generic'
   auto priv2 = []() __generic {};
   priv2();
-  auto priv3 = []() __global {}; //expected-note{{candidate function not viable: 'this' object is in address space '__private', but method expects object in address space '__global'}} //expected-note{{conversion candidate of type 'void (*)()'}}
-  priv3(); //expected-error{{no matching function for call to object of type}}
+  auto priv3 = []() __global {}; //ex pected-note{{candidate function not viable: 'this' object is in address space '__private', but method expects object in address space '__global'}} //ex pected-note{{conversion candidate of type 'void (*)()'}}
+  priv3(); //ex pected-error{{no matching function for call to object of type}}
```

Comment:
@AnastasiaStulova, I'd like to check with you if this is a bug in C++ for OpenCL compiler or not.
NOTE, this patch enables conversion to from/to unqualified types and as OpenCL qualifies all the types it should not be affected.
All OpenCL tests pass except this one test case.
Is it a bug in function pointer type inference for lambdas?

Reply:
Can you explain why you think this is a bug?

In OpenCL all local variables are deduced to be in `__private` addr space so it is the case for function object `priv3`, however it's call operator has `__global` qualifier and hence can only be used with function objects in `__global` addr space because no other addr space can implicitly be converted to `__global`. So giving an error on the call to this call operator seems correct.

Reply:
I think the issue might be due to your downstream change in `isAddressSpaceSupersetOf` that I don't entirely understand.

Reply:
`isAddressSpaceSupersetOf` allow conversion between unqualified types and types qualified with `opencl_global`, `opencl_local` and `opencl_private` address space qualifiers.
As I mentioned above it should have zero impact on OpenCL compiler, which is supposed to qualify *ALL* types with address space qualifier.


> In OpenCL all local variables are deduced to be in `__private` addr space so it is the case for function object `priv3`, however its call operator has `__global` qualifier and hence can only be used with function objects in `__global` addr space because no other addr space can implicitly be converted to `__global`. So giving an error on the call to this call operator seems correct.

It doesn't seem to work as you described if address space is deduced correctly, the mismatch will still be reported. It looks like compiler relies somehow on `isAddressSpaceSupersetOf ` in deduction process and `priv3` type is not qualified with the address space. I'll check this assumption.

Reply:
> It doesn't seem to work as you described if address space is deduced correctly, the mismatch will still be reported. 

Sorry what doesn't work as I describe?

> It looks like compiler relies somehow on `isAddressSpaceSupersetOf ` in deduction process and `priv3` type is not qualified with the address space. I'll check this assumption.

`isAddressSpaceSupersetOf ` is not used in deduction it is used in call operator overloading resolution.

Reply:
> `isAddressSpaceSupersetOf` allow conversion between unqualified types and types qualified with opencl_global, opencl_local and opencl_private address space qualifiers.

I don't see why would you allow this? 

> As I mentioned above it should have zero impact on OpenCL compiler, which is supposed to qualify ALL types with address space qualifier.

There are minor examples of undeduced addr spaces in OpenCL i.e. temporaries. Although we intend to fix those. But for now they will appear in some situations.

Reply:
I think the best documentation that described addr space qualifiers on a function is in Section 2.2.5 of
https://github.com/KhronosGroup/Khronosdotorg/blob/master/api/opencl/assets/CXX_for_OpenCL.pdf 

Lambdas are not described there yet but they are represented as callable objects in AST. So it is conceptually similar (if you look at AST representation for this example).

```
 |-DeclStmt 0x559f1520eea8 <line:34:3, col:32>
  | | `-VarDecl 0x559f1520e780 <col:3, col:31> col:8 used priv3 '__private (lambda at ../clang/test/SemaOpenCLCXX/address-space-lambda.cl:34:16)':'__private (lambda at ../clang/test/SemaOpenCLCXX/address-space-lambda.cl:34:16)' cinit
  | |   `-ImplicitCastExpr 0x559f1520ee90 <col:16, col:31> '__private (lambda at ../clang/test/SemaOpenCLCXX/address-space-lambda.cl:34:16)':'__private (lambda at ../clang/test/SemaOpenCLCXX/address-space-lambda.cl:34:16)' <NoOp>
  | |     `-LambdaExpr 0x559f1520ec90 <col:16, col:31> '(lambda at ../clang/test/SemaOpenCLCXX/address-space-lambda.cl:34:16)'
  | |       |-CXXRecordDecl 0x559f1520e840 <col:16> col:16 implicit class definition
  | |       | |-DefinitionData lambda pass_in_registers empty standard_layout trivially_copyable literal can_const_default_init
  | |       | | |-DefaultConstructor defaulted_is_constexpr
  | |       | | |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
  | |       | | |-MoveConstructor exists simple trivial needs_implicit
  | |       | | |-CopyAssignment trivial has_const_param needs_implicit implicit_has_const_param
  | |       | | |-MoveAssignment
  | |       | | `-Destructor simple irrelevant trivial
  | |       | |-CXXMethodDecl 0x559f1520e980 <col:19, col:31> col:16 constexpr operator() 'void () const __global' inline
  | |       | | `-CompoundStmt 0x559f1520ea30 <col:30, col:31>
  | |       | |-CXXConversionDecl 0x559f1520eb28 <col:16, col:31> col:16 implicit constexpr operator void (*)() 'void (*() const noexcept)()' inline
  | |       | |-CXXMethodDecl 0x559f1520ebd8 <col:16, col:31> col:16 implicit __invoke 'void ()' static inline
  | |       | `-CXXDestructorDecl 0x559f1520ecc0 <col:16> col:16 implicit referenced ~ 'void () __generic noexcept' inline default trivial
  | |       `-CompoundStmt 0x559f1520ea30 <col:30, col:31>

```

Reply:
Shouldn't the error be reported for the declaration (line 34) rather than call expression (line 35)?

Reply:
Do you know where `isAddressSpaceSupersetOf` called from for this line?

Reply:
It's not how lambda parsing works in C++. The initializer is parsed separately and there is no danger of declaring it this way unless it's being called.

I would suggest to check the comments on the review too that might help to clarify the topic:
https://reviews.llvm.org/D70242
https://reviews.llvm.org/D69938

Reply:
https://reviews.llvm.org/D69938 - the discussion started with an assumption that deduced address space for lambda object can be changed to legalize function call operator usages - https://reviews.llvm.org/D69938#1737196, but ended with an open question if it's really possible - https://reviews.llvm.org/D69938#1755709.
Later the conclusion was to use "opencl_generic" for temporaries rather than "default" - https://reviews.llvm.org/D69938#1759241. Why does current implementation use `default` address space instead of `opencl_generic`?

It's still not clear whether it's the right approach to detect an error at the function call operator use or it should be done for lambda variable declaration. The case where address space of a lambda object is changed to make use of inconsistent declaration is not covered by tests.

Reply:
Alright, I dug in. This fail basically happens because It seems that lambdas have predefined conversions to function pointer form:
```
`-LambdaExpr 0x563552014f70 <col:16, col:31> '(lambda at address-space-lambda.cl:34:16)'
  |-CXXRecordDecl 0x563552014928 <col:16> col:16 implicit class definition
  | |-DefinitionData lambda pass_in_registers empty standard_layout trivially_copyable literal can_const_default_init
  | | |-DefaultConstructor defaulted_is_constexpr
  | | |-CopyConstructor simple trivial has_const_param needs_implicit implicit_has_const_param
  | | |-MoveConstructor exists simple trivial needs_implicit
  | | |-CopyAssignment trivial has_const_param needs_implicit implicit_has_const_param
  | | |-MoveAssignment
  | | `-Destructor simple irrelevant trivial
  | |-CXXMethodDecl 0x563552014a60 <col:19, col:31> col:16 used constexpr operator() 'void () const __global' inline
  | | `-CompoundStmt 0x563552014b10 <col:30, col:31>
  | |-[[THIS ONE]] CXXConversionDecl 0x563552014e08 <col:16> col:16 implicit used constexpr operator void (*)() 'void (*() const noexcept)()' inline
  | | `-CompoundStmt 0x563552044e78 <col:16>
  | |   `-ReturnStmt 0x563552044e68 <col:16>
  | |     `-ImplicitCastExpr 0x563552044e50 <col:16> 'void (*)()' <FunctionToPointerDecay>
  | |       `-DeclRefExpr 0x563552044e30 <col:16> 'void ()' lvalue CXXMethod 0x563552014eb8 '__invoke' 'void ()'
  | |-CXXMethodDecl 0x563552014eb8 <col:16> col:16 implicit used __invoke 'void ()' static inline
  | | `-CompoundStmt 0x563552044e20 <col:16>
  | `-CXXDestructorDecl 0x563552014fa0 <col:16> col:16 implicit referenced ~ 'void () __generic noexcept' inline default trivial
  `-CompoundStmt 0x563552014b10 <col:30, col:31>

```
C++ requres it:

> The closure type for a lambda-expression with no lambda-capture has a public non-virtual non-explicit const conversion function to pointer to function having the same parameter and return types as the closure type’s function call operator. 

But as you can see from the AST dump this conversion operator converts this lambda to a poiner to a function without address space qualifiers at all, i,e, everything what this function accepts can be in **Default** address space. Whereas calling operator of such lambda has `__global` qualifier. 
After we allow conversion between between **Default** address space and others this conversion is used by compiler to make such call of this lambda valid:
```
`-CallExpr 0x563552044ec8 <line:35:3, col:9> 'void':'void'
  `-ImplicitCastExpr 0x563552044eb0 <col:3> 'void (*)()' <UserDefinedConversion>
    `-CXXMemberCallExpr 0x563552044e90 <col:3> 'void (*)()'
      `-MemberExpr 0x563552044df0 <col:3> '<bound member function type>' .operator void (*)() [[THE CONVERSION!!!!]] 0x563552014e08
        `-ImplicitCastExpr 0x563552044dd8 <col:3> 'const (lambda at address-space-lambda.cl:34:16)' lvalue <AddressSpaceConversion>
          `-DeclRefExpr 0x563552044d70 <col:3> '__private (lambda at address-space-lambda.cl:34:16)':'__private (lambda at address-space-lambda.cl:34:16)' lvalue Var 0x563552014820 'priv3' '__private (lambda

```

I'm not an expert in OpenCL and C++ for OpenCL but, if the philosophy of OpenCL address spaces is qualify ALL types with address space qualifier and if C++ for OpenCL should work like regular C++ for language features like lambdas, I think conversion from lambda object to function pointer must convert to a pointer to a function with the same address space qualifiers as operator `()` has. So, IMO this is a bug in C++ for OpenCL. @AnastasiaStulova , please provide your opinion on this.

Reply:
Submitted a bug https://bugs.llvm.org/show_bug.cgi?id=45472 .